### DISCUSSION TITLE
Broken code points fallback on MacOS: U+21E1, U+21E3

### DISCUSSION THREAD

**Discussion by lemeb** [ghostty-org/ghostty#3541] [Feature Requests, Ideas] [answered]
Hi everyone!

I believe that there's a few code points with broken fallbacks on MacOS:

## Details

* The code points are ⇡(`U+21E1`) and  ⇣(`U+21E3`). They are **not** Nerd Font glyphs, but [standard Unicode arrows](https://codepoints.net/arrows). They are relatively widely used, in particular because [Powerlevel10k uses them for displaying Git status info](https://github.com/search?q=repo%3Aromkatv%2Fpowerlevel10k+⇡&type=issues).
* On my machine, it seems like the fallback font for these code points is "PCMyungjo", but for each code point the character is not properly centered.
* As far as I can tell, none of the bundled Ghostty fonts carry these code points. (And the standard JetBrainsMono NF, which I downloaded separately, doesn't seem to have it.)
* WezTerm uses Menlo as fallback for these code points. 

## Screenshots

On this screenshot, Ghostty is on top, and WezTerm on the bottom
<img width="1208" alt="Screenshot 2024-12-27 at 19 48 10" src="https://github.com/user-attachments/assets/0dabacfb-fe76-4311-90c3-ff110d85ea72" />

## Fixes

* `font-codepoint-map = U+21E1-U+21E3=Menlo` works fine. In this font the character seems properly centered.

See screenshot:
<img width="674" alt="Screenshot 2024-12-27 at 19 57 18" src="https://github.com/user-attachments/assets/3dc8d6db-5145-497b-a7e2-44f520cfa707" />

## Setup

* New-ish M-series MBP on 15.2
* No Ghostty config

**Comment by mitchellh** [comment]
Our fallback search should be looked into here in why we're getting the wrong thing, but I'll also add this seems like a symbol we should rasterize in our sprite font (programmatically). Its simple enough and symbolic.

**Comment by lemeb** [comment]
Yes!

As an aside, I'm not sure why the fallback font changes based on the code point, because for some Menlo is selected, and for others it's PCMyungjo, which in general doesn't center characters properly.

<img width="654" alt="Screenshot 2024-12-27 at 20 05 21" src="https://github.com/user-attachments/assets/574e4adb-72bc-45c1-9b3d-40efbeacaa67" />

Here's the zsh one-liner to replicate:
```zsh
# Just for arrows code points
# See: https://codepoints.net/arrows
start=0x2190
end=0x21FF

while (( start <= end )); do
  ghostty +show-face --cp=$(printf "0x%X" $start)
  (( start++ ))
done
```

The weird thing is that deactivating PCMyungjo makes the problem go away — but I'm not sure it's a good idea to deactivate a system font like that...

**Comment by zzhaolei** [comment]
I also encountered the same problem

`echo "ヾ(☆▽☆)" "⇣5"`

<img width="607" alt="image" src="https://github.com/user-attachments/assets/37551e2c-a435-4df2-bb9a-f310023b3ea5" />
<img width="344" alt="image" src="https://github.com/user-attachments/assets/c7587b30-df58-444e-9773-f39d2e33ebc3" />

**Comment by blurgyy** [comment]
`font-codepoint-map` is working only for some codepoints on my linux machine.  I found that some codepoints just does not get mapped.  To reproduce, here is my version:

```
Ghostty 1.0.1-87bd0bb-nix

Version
  - version: 1.0.1-87bd0bb-nix
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.3
    runtime    : 4.16.3
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
  - libX11     : enabled
```

The config content:

```
font-family = "Iosevka Fixed"
font-codepoint-map = U+E0A0 = Symbols Nerd Font
font-codepoint-map = U+0030 = Apple Color Emoji
```

And run the following 3 commands, observe that `e0a0` fails to be mapped:

```bash
$ ghostty +show-face --cp=0x0030  # this is good
U+30 « 0 » found in face “Apple Color Emoji”.

$ ghostty +show-face --cp=0x0031  # this is also good
U+31 « 1 » found in face “Iosevka Fixed”.

$ ghostty +show-face --cp=0xe0a0  # <-- this is not mapped
U+E0A0 «  » found in face “Iosevka Fixed”.
```

---

Edit: to be clear, the result of the last command should be "Symbols Nerd Font" as the config specifies.  I have blacklisted the codepoint range `0xe000` to `0xf8ff` from "Iosevka Fixed" in my fontconfig, and here is fontconfig matching the codepoint to the correct font:

```bash
$ fc-match :charset=0xe0a0:family=monospace
SymbolsNerdFont-Regular.ttf: "Symbols Nerd Font" "Regular"
```

**Comment by danielwe** [comment, marked as answer]
Flying by and noticing that #8693 was actually a duplicate of this. So if anyone with privileges sees this, I think this can be closed now (see #9189).
